# DS Practicum 1

## Question 1

**Problem:**

## Question 2

**Loading Data**

In [3]:
import pandas as pd

# Load the CSV with the correct filename
df = pd.read_csv("2019_ONTIME_REPORTING_FSW-1.csv")

# View the first 5 rows
df.head()

FL_DATE CARRIER_CODE TAIL_NUM  FL_NUM ORIGIN ORIGIN_ST DEST DEST_ST  \
0  2019-10-01           AA   N916NN    2311    TUS        AZ  ORD      IL   
1  2019-10-01           AA   N733UW    2315    PHX        AZ  DEN      CO   
2  2019-10-01           AA   N140AN    2318    DFW        TX  LAX      CA   
3  2019-10-01           AA   N925AN    2325    SNA        CA  DFW      TX   
4  2019-10-01           AA   N143AN    2328    ATL        GA  LAX      CA   

   DEP_TIME  DEP_DELAY  ARR_TIME  ARR_DELAY  ELAPSED_TIME  DISTANCE  
0     828.0        0.0    1353.0        0.0         205.0      1437  
1    1907.0        0.0    2159.0        0.0         112.0       602  
2    1904.0      104.0    2016.0      101.0         192.0      1235  
3    1729.0        4.0    2215.0        3.0         166.0      1205  
4     656.0        0.0     841.0        0.0         285.0      1947

**Inspecting Data and Converting to Suitable Types**

In [4]:
# inspect the dimensions
print(f"There are {df.shape[0]} rows and {df.shape[1]} columns.")

# inspect data types
df.info()

There are 1897503 rows and 14 columns.
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1897503 entries, 0 to 1897502
Data columns (total 14 columns):
 #   Column        Dtype  
---  ------        -----  
 0   FL_DATE       object 
 1   CARRIER_CODE  object 
 2   TAIL_NUM      object 
 3   FL_NUM        int64  
 4   ORIGIN        object 
 5   ORIGIN_ST     object 
 6   DEST          object 
 7   DEST_ST       object 
 8   DEP_TIME      float64
 9   DEP_DELAY     float64
 10  ARR_TIME      float64
 11  ARR_DELAY     float64
 12  ELAPSED_TIME  float64
 13  DISTANCE      int64  
dtypes: float64(5), int64(2), object(7)
memory usage: 202.7+ MB


**Dimensions** <br></br>There are 1897503 rows and 14 columns of data.  

**Changes in Data Types**
<br></br>
CARRIER_CODE, TAIL_NUM, FL_NUM, ORIGIN, ORIGIN_ST, DEST, and DEST_ST will be changed to category variables because they represent labels.  FL_DATE will be changed to datetime because it represents a date.  DEP_TIME, DEP_DELAY, ARR_TIME, ARR_DELAY, and ELAPSED_TIME will be changed to integers as well.

In [5]:
# change FL_DATE to datetime
df["FL_DATE"] = pd.to_datetime(df["FL_DATE"])

# change CARRIER_CODE, TAIL_NUM, FL_NUM, ORIGIN, ORIGIN_ST, DEST, DEST_ST to
# categorical variables
df["FL_NUM"] = df["FL_NUM"].astype("category")
df["CARRIER_CODE"] = df["CARRIER_CODE"].astype("category")
df["TAIL_NUM"] = df["TAIL_NUM"].astype("category")
df["ORIGIN"] = df["ORIGIN"].astype("category")
df["ORIGIN_ST"] = df["ORIGIN_ST"].astype("category")
df["DEST"] = df["DEST"].astype("category")
df["DEST_ST"] = df["DEST_ST"].astype("category")

# change DEP_TIME, DEP_DELAY, ARR_TIME, ARR_DELAY, ELAPSED_TIME to Int64
df["DEP_TIME"] = df["DEP_TIME"].astype("Int64")
df["DEP_DELAY"] = df["DEP_DELAY"].astype("Int64")
df["ARR_TIME"] = df["ARR_TIME"].astype("Int64")
df["ARR_DELAY"] = df["ARR_DELAY"].astype("Int64")
df["ELAPSED_TIME"] = df["ELAPSED_TIME"].astype("Int64")

# verify changes 
df.dtypes

FL_DATE         datetime64[ns]
CARRIER_CODE          category
TAIL_NUM              category
FL_NUM                category
ORIGIN                category
ORIGIN_ST             category
DEST                  category
DEST_ST               category
DEP_TIME                 Int64
DEP_DELAY                Int64
ARR_TIME                 Int64
ARR_DELAY                Int64
ELAPSED_TIME             Int64
DISTANCE                 int64
dtype: object

In [6]:
# verify changes by checking first 5 rows
df.head()

FL_DATE CARRIER_CODE TAIL_NUM FL_NUM ORIGIN ORIGIN_ST DEST DEST_ST  \
0 2019-10-01           AA   N916NN   2311    TUS        AZ  ORD      IL   
1 2019-10-01           AA   N733UW   2315    PHX        AZ  DEN      CO   
2 2019-10-01           AA   N140AN   2318    DFW        TX  LAX      CA   
3 2019-10-01           AA   N925AN   2325    SNA        CA  DFW      TX   
4 2019-10-01           AA   N143AN   2328    ATL        GA  LAX      CA   

   DEP_TIME  DEP_DELAY  ARR_TIME  ARR_DELAY  ELAPSED_TIME  DISTANCE  
0       828          0      1353          0           205      1437  
1      1907          0      2159          0           112       602  
2      1904        104      2016        101           192      1235  
3      1729          4      2215          3           166      1205  
4       656          0       841          0           285      1947

In [7]:
# percentage of missing values
df.isnull().mean()*100

FL_DATE         0.000000
CARRIER_CODE    0.000000
TAIL_NUM        0.252490
FL_NUM          0.000000
ORIGIN          0.000000
ORIGIN_ST       0.000000
DEST            0.000000
DEST_ST         0.000000
DEP_TIME        1.407481
DEP_DELAY       1.407903
ARR_TIME        1.488482
ARR_DELAY       1.680314
ELAPSED_TIME    1.680314
DISTANCE        0.000000
dtype: float64

In [8]:
# distribution of missing values in rows
df.isnull().sum(axis=1).sort_values(ascending=False).value_counts()

0    1865619
5      21918
6       4790
2       3640
3       1529
4          7
Name: count, dtype: int64

**Missing Values**<br></br>Based on this information, missing data is not widespread since only a small percentage of rows have missing data.  Out of the rows with missing data, most have 5 or 6 missing columns.  Since the columns with missing values are related to arrival, departure, and delay times and this is important to the data science question on air traffic in part 1, these rows should be dropped.

In [9]:
# remove rows with null values
df = df.dropna()

# verify changes (only rows with 0 null values retained)
df.isnull().sum(axis=1).sort_values(ascending=False).value_counts()

0    1865619
Name: count, dtype: int64

**Invalid Data**

Invalid times of 2400 for DEP_TIME were corrected by changing them to 0 and updating the FL_DATE by 1 day.  Rows with times of 2400 were not excluded since the data is not necessarily incorrect, just inconsistent since 2400 means 24 hours later or the next day.  This means the flight actually departed at midnight the next day.

In [ ]:
# determine number of rows with invalid DEP_TIME
invalid_dep_time_df = df[df["DEP_TIME"] == 2400]
print(f"{invalid_dep_time_df.shape[0]} rows with invalid DEP_TIME values")
# display rows with invalid departure times to verify later
invalid_dep_time_df.head()

241 rows with invalid DEP_TIME values


FL_DATE CARRIER_CODE TAIL_NUM FL_NUM ORIGIN ORIGIN_ST DEST DEST_ST  \
34864 2019-10-07           DL   N383DZ    487    LAS        NV  ATL      GA   
40416 2019-10-08           AA   N825NN   1542    PHX        AZ  DCA      VA   
50938 2019-10-10           DL   N841MH   1362    LAX        CA  JFK      NY   
65478 2019-10-13           NK   N627NK    456    LAS        NV  ORD      IL   
68336 2019-10-14           DL   N844MH   1362    LAX        CA  JFK      NY   

       DEP_TIME  DEP_DELAY  ARR_TIME  ARR_DELAY  ELAPSED_TIME  DISTANCE  
34864      2400          1       635          0           215      1747  
40416      2400          0       707          0           247      1979  
50938      2400          5       817          0           317      2475  
65478      2400          0       517          0           197      1514  
68336      2400          5       803          0           303      2475

There are 241 rows with invalid arrival times.

In [37]:
# update rows with invalid DEP_TIME
dep_time_filter = df["DEP_TIME"] == 2400
df.loc[dep_time_filter, "DEP_TIME"] = 0
df.loc[dep_time_filter, "FL_DATE"] = df.loc[dep_time_filter, "FL_DATE"] + pd.DateOffset(days=1)

# verify changes
new_dep_time_df = df[df["DEP_TIME"] == 2400]
print(f"{new_dep_time_df.shape[0]} rows with invalid DEP_TIME values")

0 rows with invalid DEP_TIME values


## Question 3

## Question 4

## Question 5

## Question 6

## Question 7